src/plugins/data_views/server/rest_api_routes/internal/has_es_data.ts

```diff
@@ -7,34 +7,122 @@
-    name: patterns.concat(crossClusterPatterns),
-    allow_no_indices: true,
-    ignore_unavailable: true,
-  });
+export const createHandler =
+  (parentLogger: Logger, hasEsDataTimeout: number): Handler =>
+  async (ctx, _, res) => {
+    const logger = parentLogger.get('hasEsData');
+    const core = await ctx.core;
+    const elasticsearchClient = core.elasticsearch.client.asCurrentUser;
+    const commonParams: Omit<HasEsDataParams, 'matchPatterns' | 'timeoutReason'> = {
+      elasticsearchClient,
+      logger,
+      res,
+      hasEsDataTimeout,
+    };
 
-  const hasEsData = !!Object.values(response).find((cluster) => cluster.matching_indices);
+    const localDataResponse = await hasEsData({
+      ...commonParams,
+      matchPatterns: patterns,
+      timeoutReason: HasEsDataFailureReason.localDataTimeout,
+    });
 
-  return res.ok({ body: { hasEsData } });
+    if (localDataResponse) {
+      return localDataResponse;
+    }
+
+    const remoteDataResponse = await hasEsData({
+      ...commonParams,
+      matchPatterns: crossClusterPatterns,
+      timeoutReason: HasEsDataFailureReason.remoteDataTimeout,
+    });
+
+    if (remoteDataResponse) {
+      return remoteDataResponse;
+    }
+
+    return res.ok({ body: { hasEsData: false } });
+  };
+
+interface HasEsDataParams {
+  elasticsearchClient: ElasticsearchClient;
+  logger: Logger;
+  res: KibanaResponseFactory;
+  matchPatterns: string[];
+  hasEsDataTimeout: number;
+  timeoutReason: HasEsDataFailureReason;
+}
+
+const timeoutMessage = 'Timeout while checking for Elasticsearch data';
+const errorMessage = 'Error while checking for Elasticsearch data';
+
+const hasEsData = async ({
+  elasticsearchClient,
+  logger,
+  res,
+  matchPatterns,
+  hasEsDataTimeout,
+  timeoutReason,
+}: HasEsDataParams) => {
+  try {
+    const response = await elasticsearchClient.indices.resolveCluster(
+      {
+        name: matchPatterns,
+        allow_no_indices: true,
+        ignore_unavailable: true,
+      },
+      { requestTimeout: hasEsDataTimeout === 0 ? undefined : hasEsDataTimeout }
+    );
+
+    const hasData = Object.values(response).some((cluster) => cluster.matching_indices);
+
+    if (hasData) {
+      return res.ok({ body: { hasEsData: true } });
+    }
+  } catch (e) {
+    if (e.name === 'TimeoutError') {
+      const warningMessage =
+        `${timeoutMessage}: ${timeoutReason}. Current timeout value is ${hasEsDataTimeout}ms. ` +
+        `Use "data_views.hasEsDataTimeout" in kibana.yml to change it, or set to 0 to disable timeouts.`;
+
+      logger.warn(warningMessage);
+
+      return res.badRequest({
+        body: {
+          message: timeoutMessage,
+          attributes: { failureReason: timeoutReason },
+        },
+      });
+    }
+
+    logger.error(e);
+
+    return res.badRequest({
+      body: {
+        message: errorMessage,
+        attributes: { failureReason: HasEsDataFailureReason.unknown },
+      },
+    });
```

Comment:
This one I'm not sure what to do... We can probably leave as is. Are there any known cases we might fail here? If so, we might want to check `e.meta.statusCode` and use it.

Reply:
Yeah, also not a good case for 400. And nope, no known cases... Which makes me realize it probably makes sense to just return a 500 here since it's unexpected. I think this is good enough for the client, and we log the underlying error if needed for further investigation. Updated here: https://github.com/elastic/kibana/pull/200476/commits/f130736e12d6c1fe546b13f1a03f0b75473a10b3.